cub/cub/thread/thread_sort.cuh

```diff
@@ -37,25 +37,79 @@
 #  pragma system_header
 #endif // no system header
 
+#include <cub/thread/thread_operators.cuh>
 #include <cub/util_ptx.cuh>
 #include <cub/util_type.cuh>
 
+#include <cuda/std/cstdint>
 #include <cuda/std/type_traits>
 
 CUB_NAMESPACE_BEGIN
 
+template <typename IntT>
+_CCCL_HOST_DEVICE _CCCL_FORCEINLINE constexpr IntT NetworkDegree(IntT n, IntT m = IntT{1})
+{
+  return (m < n) ? NetworkDegree(n, m * 2) + 1 : 0;
+}
 
 template <typename T>
-_CCCL_DEVICE _CCCL_FORCEINLINE void Swap(T &lhs, T &rhs)
+_CCCL_DEVICE _CCCL_FORCEINLINE void Swap(T& lhs, T& rhs)
 {
   T temp = lhs;
   lhs    = rhs;
   rhs    = temp;
 }
 
+template <typename KeyT>
+_CCCL_DEVICE _CCCL_FORCEINLINE void CompareSwapMinMaxAsc(KeyT& key_lhs, KeyT& key_rhs)
+{
+  KeyT pair_min = (cub::min)(key_lhs, key_rhs);
+  KeyT pair_max = (cub::max)(key_lhs, key_rhs);
+  key_lhs       = pair_min;
+  key_rhs       = pair_max;
+}
+
+template <typename KeyT, typename ValueT, typename CompareOp>
+_CCCL_DEVICE _CCCL_FORCEINLINE void
+CompareSwap(KeyT& key_lhs, KeyT& key_rhs, ValueT& item_lhs, ValueT& item_rhs, CompareOp compare_op)
+{
+  constexpr bool KEYS_ONLY = ::cuda::std::is_same<ValueT, NullType>::value;
+
+  if (compare_op(key_rhs, key_lhs))
+  {
+    Swap(key_lhs, key_rhs);
+    CUB_IF_CONSTEXPR(!KEYS_ONLY)
+    {
+      Swap(item_lhs, item_rhs);
+    }
+  }
+}
+
+#define SPECIALIZE_SORT_ASC(T)                                                            \
+  template <>                                                                             \
+  _CCCL_DEVICE _CCCL_FORCEINLINE void CompareSwap(                                        \
+    T& key_lhs, T& key_rhs, NullType& item_lhs, NullType& item_rhs, cub::Less compare_op) \
+  {                                                                                       \
+    CompareSwapMinMaxAsc(key_lhs, key_rhs);                                               \
+  }
+
+#define SPECIALIZE_SORT_DESC(T)                                                              \
+  template <>                                                                                \
+  _CCCL_DEVICE _CCCL_FORCEINLINE void CompareSwap(                                           \
+    T& key_lhs, T& key_rhs, NullType& item_lhs, NullType& item_rhs, cub::Greater compare_op) \
+  {                                                                                          \
+    CompareSwapMinMaxAsc(key_rhs, key_lhs);                                                  \
+  }
+
+SPECIALIZE_SORT_ASC(::cuda::std::int32_t)
```

Comment:
For arithmetic types and keys only, using min and max takes two instructions per swap, whereas the conditional version takes 3 (1 set predicate, 2 selections). For integers, the compiler does the optimization automatically, so this specialization is not strictly required, but for float32 it can't do the optimization because the behavior is slightly different with special cases like NaN.

Regarding NaN, the conditional version would not produce a sorted array if there are any NaNs, because comparisons with NaN always evaluate to false, breaking the rules of strict weak ordering:

```
IN:  { 5, 4, NaN, 3, 2, 1 }
OUT: { 4, 5, NaN, 1, 2, 3 }
```

With the min/max version, afaik if one input of CompareSwap is NaN it will duplicate the other, so a possible output of the sort would be:

```
IN:  { 5, 4, NaN, 3, 2, 1 }
OUT: { 1 2 3 3 4 5 }
```

The point is that I regard NaNs in the array as invalid inputs and prefer to use the fast implementation with 2 instructions instead of 3.